### PR TITLE
Avro test

### PR DESCRIPTION
**Pull request by RoriCremer**

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohort.java [hunk 1]
```diff
@@ -36,7 +36,7 @@
     @Argument(
             fullName = "cohort-extract-table",
             doc = "Fully qualified name of the table where the cohort data exists (already subsetted)",
-            optional = false
+            optional = true
```
Comment (mmorgantaylor) [inline src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohort.java:39]: is this change needed for the case where you're using an avro file directly rather than pulling from bq? and did we come up with any use cases for this other than ease of testing?

File: src/test/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/AnVIL_WGS_100_exported_cohort/sample_list [hunk 2]
```diff
@@ -0,0 +1,100 @@
+30,HG00561
```
Comment (mmorgantaylor) [inline src/test/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/AnVIL_WGS_100_exported_cohort/sample_list:1]: probably could use just a few samples, and use a subsetted interval list, so that the test files can be small. there's a small interval list in the CreateVariantIngestFiles integration test resources, if that helps!

File: src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohort.java (hunk 1 above)
Comment (mmorgantaylor) [reply src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohort.java:39]: what if we keep this as `optional = false`, and then change the logic in ExtractCohortEngine from `if (cohortTableRef != null){` to `if (cohortAvroFileName == null)` ? 

then in your test you can provide a fake cohort extract table name, but it'll still read from your avro file.

File: src/test/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohortEngineTest.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import org.broadinstitute.hellbender.testutils.IntegrationTestSpec;
+import org.testng.annotations.Test;
+
+import java.io.File;
+
+
+public class ExtractCohortEngineTest extends CommandLineProgramTest {
+
+  private static final String cohortAvroFileName = // is it okay to just grab one?
+      "src/test/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/AnVIL_WGS_100_exported_cohort/000000000000";
```
Comment (mmorgantaylor) [inline src/test/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohortEngineTest.java:14]: can use `getToolTestDataDir()` to get the path

File: src/test/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/AnVIL_WGS_100_exported_cohort/sample_list (hunk 2 above)
Comment (kcibul) [reply src/test/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/AnVIL_WGS_100_exported_cohort/sample_list]: Yes please!  Just take a few ( < 5 ) samples and just a few sites (100s)

File: src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohort.java [hunk 4]
```diff
@@ -145,6 +142,12 @@ protected void onStartup() {
             throw new UserException("min-location and max-location should not be used together with intervals (-L).");
         }
 
+        // if there is a avro file, the BQ specific parameters are unnecessary,
+        // but they all are required if there is no avro file
+        if (cohortAvroFileName == null && (projectID == null || cohortTable == null)) {
+            throw new UserException("a project id and cohort table are required if no avro file is provided");
```
Comment (mmorgantaylor) [inline src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohort.java:148]: super small, but for ease of the user, could you provide the argument flags corresponding to project id and cohort table? e.g. `Project id (--project-id) and cohort table (--cohort-extract-table) are required if no avro file is provided.`

**Comment by kcibul** [comment]
One concern I have is the maintainability of the test (having been burned by this in other places myself).  When we add a new output field, etc we need a very easy way to update/generate these results.  At the very least some instructions would be helpful (and imagine someone to follow those as part of a PR)

File: src/test/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohortTest.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import org.broadinstitute.hellbender.testutils.ArgumentsBuilder;
+import org.broadinstitute.hellbender.testutils.IntegrationTestSpec;
+import org.testng.annotations.Test;
+
+import java.io.File;
+
+
+class ExtractCohortTest extends CommandLineProgramTest {
+  private final String prefix = getToolTestDataDir();
+  private final String cohortAvroFileName = prefix +"000000000000.avro";
```
Comment (mmorgantaylor) [inline src/test/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohortTest.java:13]: this file is 22.8 MB - can we make it smaller?

File: src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohort.java [hunk 6]
```diff
@@ -52,13 +47,15 @@
...
-            optional = false
+            mutex={"cohort-avro-file-name"},
+            optional = true
     )
     private String cohortTable = null;
 
     @Argument(
             fullName = "cohort-avro-file-name",
             doc = "Path of the cohort avro file",
+            mutex={"cohort-extract-table"},
```
Comment (mmorgantaylor) [inline src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohort.java:58]: i'm so sorry to be this person but can you put spaces before and after the `=` on all these mutex sets so it's consistent?

File: src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohort.java (hunk 1 above)
Comment (mmorgantaylor) [reply src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohort.java:51]: these comments are moot now

**Review by mmorgantaylor — APPROVED** [review]
Looks good! Some small comments, and I think we shouldn't have a 23 MB test file, but other than that 👍

File: src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohort.java [hunk 7]
```diff
@@ -145,6 +142,13 @@ protected void onStartup() {
             throw new UserException("min-location and max-location should not be used together with intervals (-L).");
         }
 
+        // if there is a avro file, the BQ specific parameters are unnecessary,
+        // but they all are required if there is no avro file
+        if (cohortAvroFileName == null && (projectID == null || cohortTable == null)) {
+            throw new UserException("Project id (--project-id) " +
+                "and cohort table (--cohort-extract-table) are required if no avro file is provided.");
```
Comment (mmorgantaylor) [inline src/main/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohort.java:149]: ❤️

File: src/test/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohortTest.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+class ExtractCohortTest extends CommandLineProgramTest {
+  private final String prefix = getToolTestDataDir();
+  private final String cohortAvroFileName = prefix +"chr20_subset_3_samples.avro";
+  private final String sampleFile = prefix +"sample_list";
+  private final String outputFileName = prefix +"output.vcf";
+
+  @Test
+  public void testFinalVCFfromAvro() throws Exception {
+    // To create the expected output file--create a temp table in BQ with the folowing query
```
Comment (kcibul) [inline src/test/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohortTest.java:19]: awesome

File: src/test/resources/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohort/sample_list [hunk 9]
```diff
@@ -0,0 +1,3 @@
+2,HG00408
```
Comment (kcibul) [inline src/test/resources/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohort/sample_list:1]: nit: can this be sorted?

File: src/test/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohortTest.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+        .add("R", hg38Reference)
+        .add("O", outputFileName)
+        .add("local-sort-max-records-in-ram", 10000000)
+        .add("cohort-avro-file-name", cohortAvroFileName)
+        .add("sample-file", sampleFile);
+
+    runCommandLine(args);
+
+    final File expectedVCF = getTestFile("expected.vcf");
+    final File outputVCF = getTestFile("output.vcf");
```
Comment (kcibul) [inline src/test/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohortTest.java:39]: Is this going to get the `output.vcf` that you checked in as part of this commit?  It should be the output of the run above (which I think won't live in the same place as getTestFile().  You can tell for sure by removing `src/test/resources/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohort/output.vcf` from your PR (which I think shouldn't be commited anyway).

**Review by kcibul — APPROVED** [review]
Looks great Rori!   Just make sure that after removing output.vcf from your commit/PR that the test still pass.

File: src/test/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohortTest.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+        .add("ref-version", 38)
+        .add("query-mode", "LOCAL_SORT")
+        .add("R", hg38Reference)
+        .add("O", outputVCF.getAbsolutePath())
+        .add("local-sort-max-records-in-ram", 10000000)
+        .add("cohort-avro-file-name", cohortAvroFileName)
+        .add("sample-file", sampleFile);
+
+    runCommandLine(args);
+    IntegrationTestSpec.assertEqualTextFiles(outputVCF, expectedVCF);
```
Comment (mmorgantaylor) [inline src/test/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohortTest.java:40]: do you also check the index file?

Comment (RoriCremer) [reply src/test/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohortTest.java:40]: this feels like it would be outside the scope of the test no? a GATK issue?

Comment (mmorgantaylor) [reply src/test/java/org/broadinstitute/hellbender/tools/variantdb/nextgen/ExtractCohortTest.java:40]: yeah, totally - i just wasn't sure why you were checking in the expected index file in that case
